src/shutdown.rs

```diff
@@ -0,0 +1,253 @@
+        }
+    }
+
+    /// Creates the necessary Triggers and Tripwires for coordinating shutdown of this Source and
+    /// stores them as needed.  Returns the ShutdownSignal for this Source as well as a Tripwire
+    /// that will be notified if the Source should be forcibly shut down.
+    pub fn register_source(&mut self, name: &str) -> (ShutdownSignal, Tripwire) {
+        let (shutdown_begun_trigger, shutdown_begun_tripwire) = Tripwire::new();
+        let (force_shutdown_trigger, force_shutdown_tripwire) = Tripwire::new();
+        let (shutdown_complete_trigger, shutdown_complete_tripwire) = Tripwire::new();
+
+        self.shutdown_begun_triggers
+            .insert(name.to_string(), shutdown_begun_trigger);
+        self.shutdown_force_triggers
+            .insert(name.to_string(), force_shutdown_trigger);
+        self.shutdown_complete_tripwires
+            .insert(name.to_string(), shutdown_complete_tripwire);
+
+        let shutdown_signal =
+            ShutdownSignal::new(shutdown_begun_tripwire, shutdown_complete_trigger);
+
+        (shutdown_signal, force_shutdown_tripwire)
+    }
+
+    /// Takes ownership of all internal state for the given source from another ShutdownCoordinator.
+    pub fn takeover_source(&mut self, name: &str, other: &mut Self) {
+        self.shutdown_begun_triggers.insert(
+            name.to_string(),
+            other.shutdown_begun_triggers.remove(name).unwrap(),
+        );
+
+        self.shutdown_force_triggers.insert(
+            name.to_string(),
+            other.shutdown_force_triggers.remove(name).unwrap(),
+        );
+
+        self.shutdown_complete_tripwires.insert(
+            name.to_string(),
+            other.shutdown_complete_tripwires.remove(name).unwrap(),
+        );
+    }
+
+    pub fn shutdown_source_begin(&mut self, name: &str) {
+        self.shutdown_begun_triggers.remove(name).unwrap().cancel();
+    }
+
+    /// Waits for the source to shut down until the deadline.  If the source does not
+    /// notify the shutdown_complete_tripwire for this source before the dealine, then signals
+    /// the shutdown_force_trigger for this source to force it to shut down.  Returns whether
+    /// or not the source shutdown gracefully.
+    // TODO: The timing and reporting logic is very similar to the logic in
+    // `RunningTopology::stop()`. Once `RunningTopology::stop()` has been updated to utilize the
+    // ShutdownCoordinator, see if some of this logic can be de-duped.
+    pub fn shutdown_source_end<'a>(
+        &mut self,
+        rt: &mut runtime::Runtime,
+        name: String,
+        deadline: Instant,
+    ) -> bool {
+        let name2 = name.clone();
+        let name3 = name.clone();
+        let shutdown_complete_tripwire = self.shutdown_complete_tripwires.remove(&name).unwrap();
+        let shutdown_force_trigger = self.shutdown_force_triggers.remove(&name).unwrap();
+
+        let success = shutdown_complete_tripwire.map(move |_| {
```

Comment:
do we want to assert here that `true` is returned?

Reply:
Not sure I understand what you're asking.  The line this comment is on is creating a Future<(),()>.  If this future is resolved before the timeout future, then that means the Source successfully shut down.  We return true or false from this function based on which Future resolves first.  Callers of `shutdown_source_end` can assert that the return value is true where it makes sense to.